src/System.Net.Http/src/System/Net/Http/Unix/CurlCallbacks.cs

```diff
@@ -0,0 +1,357 @@
+            (Interop.libcurl.curl_unsafe_write_callback) CurlReceiveBodyCallback;
+        private static Interop.libcurl.curl_readwrite_callback s_sendCallback =
+            (Interop.libcurl.curl_readwrite_callback) CurlSendCallback;
+        private static Interop.libcurl.curl_ioctl_callback s_sendIoCtlCallback =
+            (Interop.libcurl.curl_ioctl_callback) CurlSendIoCtlCallback;
+        private static Interop.libcurl.curl_socket_callback s_socketCallback =
+            (Interop.libcurl.curl_socket_callback) CurlSocketCallback;
+        private static Interop.libcurl.curl_multi_timer_callback s_multiTimerCallback =
+            (Interop.libcurl.curl_multi_timer_callback) CurlMultiTimerCallback;
+
+        private static size_t CurlReceiveHeadersCallback(
+            IntPtr buffer,
+            size_t size,
+            size_t nitems,
+            IntPtr context)
+        {
+            return ExecuteCallback(buffer, size, nitems, context,
+                (ptr, totalSize, state) =>
+                {
+                    state.CancellationToken.ThrowIfCancellationRequested();
+
+                    // The callback is invoked once per header and multi-line
+                    // headers get merged into a single line
+                    string responseHeader = Marshal.PtrToStringAnsi(ptr).Trim();
+                    HttpResponseMessage response = state.ResponseMessage;
+
+                    // TODO: Understand scenarios where multiple sets of headers are received
+                    if (!TryParseStatusLine(response, responseHeader))
+                    {
+                        int colonIndex = responseHeader.IndexOf(':');
+
+                        // Skip malformed header lines that are missing the colon character.
+                        if (colonIndex > 0)
+                        {
+                            string headerName = responseHeader.Substring(0, colonIndex);
+                            string headerValue = responseHeader.Substring(colonIndex + 1);
+
+                            if (!response.Headers.TryAddWithoutValidation(headerName, headerValue))
+                            {
+                                // TODO: Skip compression related headers
+                                response.Content.Headers.TryAddWithoutValidation(headerName, headerValue);
+                            }
+                        }
+                    }
+                    return totalSize;
+                });
+        }
+
+        private unsafe static size_t CurlReceiveBodyCallback(
+            byte* buffer,
+            size_t size,
+            size_t nitems,
+            IntPtr context)
+        {
+            size *= nitems;
+            GCHandle gch = GCHandle.FromIntPtr(context);
+            var state = (RequestCompletionSource)gch.Target;
+
+            try
+            {
+                // Set task completion after all headers have been received
+                state.TrySetResult(state.ResponseMessage);
```

Comment:
What happens at this point if TrySetResult returns false because the task is already completed, and it was completed as Canceled or Faulted, such that there will never be a reader.  Is the subsequent code that proceeds to call WaitAndSignalReaders still ok?